pyop2/host.py

```diff
@@ -133,15 +135,48 @@ def c_ind_data(self, idx, i, j=0, is_top=False, layers=1, offset=None):
              'off_add': ' + %d' % offset if not is_top and offset is not None else ''}
 
-    def c_ind_data_xtr(self, idx, i, j=0, is_top=False, layers=1):
-        return "%(name)s + (xtr_%(map_name)s[%(idx)s]%(top)s%(offset)s)*%(dim)s%(off)s" % \
-            {'name': self.c_arg_name(i),
-             'map_name': self.c_map_name(i, 0),
-             'idx': idx,
-             'top': ' + start_layer' if is_top else '',
-             'dim': 1 if self._flatten else str(self.data[i].cdim),
-             'off': ' + %d' % j if j else '',
-             'offset': ' * _'+self.c_offset_name(i, 0)+'['+idx+']' if is_top else ''}
+    def ast_ind_data(self, idx, i, j=0, is_top=False, layers=1, offset=None):
+        index = ast.Prod(ast.Symbol("i"), ast.Symbol(self.map.split[i].arity))
+        index = ast.Sum(index, ast.Symbol(idx))
+        tail = ast.Symbol(self.c_map_name(i, 0), rank=(index,))
+        if is_top:
+            tail = ast.Sum(tail, ast.Symbol("start_layer"))
+            if offset:
+                tail = ast.Prod(tail, ast.Symbol(offset))
+        if not is_top and offset:
+            tail = ast.Sum(tail, offset)
+        tail = ast.Prod(tail, ast.Symbol(self.data[i].cdim))
+        if j:
+            tail = ast.Sum(tail, ast.Symbol(j))
+        return ast.Sum(ast.Symbol(self.c_arg_name(i)), tail)
+        # return "%(name)s + (%(map_name)s[i * %(arity)s + %(idx)s]%(top)s%(off_mul)s%(off_add)s)* %(dim)s%(off)s" % \
+        #     {'name': self.c_arg_name(i),
+        #      'map_name': self.c_map_name(i, 0),
+        #      'arity': self.map.split[i].arity,
+        #      'idx': idx,
+        #      'top': ' + start_layer' if is_top else '',
+        #      'dim': self.data[i].cdim,
+        #      'off': ' + %d' % j if j else '',
+        #      'off_mul': ' * %d' % offset if is_top and offset is not None else '',
+        #      'off_add': ' + %d' % offset if not is_top and offset is not None else ''}
+
+    def ast_ind_data_xtr(self, idx, i, j=0, is_top=False, layers=1):
+        tail = ast.Symbol("xtr_" + self.c_map_name(i, 0), rank=(idx,))
+        if is_top:
+            tail = ast.Sum(tail, ast.Symbol(start_layer))
+            tail = ast.Prod(tail, ast.Symbol('_'+self.c_offset_name(i, 0), rank=(idx,)))
+        tail = ast.Prod(tail, ast.Symbol(1 if self._flatten else self.data[i].cdim))
+        if j:
+            tail = ast.Sum(tail, ast.Symbol(j))
+        return ast.Sum(ast.Symbol(self.c_arg_name(i)), tail)
+        # return "%(name)s + (xtr_%(map_name)s[%(idx)s]%(top)s%(offset)s)*%(dim)s%(off)s" % \
+        #     {'name': self.c_arg_name(i),
+        #      'map_name': self.c_map_name(i, 0),
+        #      'idx': idx,
+        #      'top': ' + start_layer' if is_top else '',
+        #      'dim': 1 if self._flatten else str(self.data[i].cdim),
+        #      'off': ' + %d' % j if j else '',
+        #      'offset': ' * _'+self.c_offset_name(i, 0)+'['+idx+']' if is_top else ''}
```

Comment:
I've left this and the other comments in on purpose because it's easier to understand what gets constructed using the AST. If I take this out I would have to replace it with a block of text explaining this in words. This would take longer and it would be less clear what the output should be.